wsr-tic-tac-toe/src/redux/reducers/index.ts

```diff
@@ -0,0 +1,137 @@
+) {
+    switch (action.type) {
+        case CLEAR_WINNING_SCENARIOS: {
+            return initialWinningScenariosState; // return initialWinningScenariosState
+        }
+        case GENERATE_WINNING_SCENARIOS: {
+            let boardSize = action.payload;
+            let WSh: Array<Set<number>> = []; // array of horizontal winning sets
+            let WSv: Array<Set<number>> = []; // array of vertical winning sets
+            for (let i: number = 0; i < boardSize; i++) {
+                let tempSet = new Set<number>();
+                for (let j: number = 0; j < boardSize; j++) {
+                    tempSet.add(boardSize * i + j);
+                }
+                WSh = [...WSh, tempSet];
+            }
+
+            for (let i: number = 0; i < boardSize; i++) {
+                let tempSet = new Set<number>();
+                for (let j: number = 0; j < boardSize; j++) {
+                    tempSet.add(boardSize * j + i);
+                }
+                WSv = [...WSv, tempSet];
+            }
+
+            let diagonalSet1 = new Set();
+            let diagonalSet2 = new Set();
+            for (let i = 0; i < boardSize; i++) {
+                diagonalSet1.add(i * boardSize + i);
+                diagonalSet2.add((i + 1) * (boardSize - 1));
+            }
+            let winningScenarios = [];
+            winningScenarios = [
+                ...WSh.map((set) => Array.from(set)),
+                ...WSv.map((set) => Array.from(set)),
+                Array.from(diagonalSet1),
+                Array.from(diagonalSet2),
+            ];
+            return [...winningScenarios];
+        }
+        default:
+            return state;
+    }
+}
+
+const initialBoardState: BoardState = {
+    boardSize: 0,
+    created: false,
+    squares: [],
+    turnOrder: 'player1',
+    gameState: GAME_STATE.ONGOING
+};
+
+export function boardReducer(state = initialBoardState, action: BoardStateActionTypes) {
+    switch (action.type) {
+        case CREATE_BOARD: {
+            let boardSize = action.payload;
+            let initializedSquares: SquareState[] = [];
+            // let squares: SquareState[][] = [];
+            for (let i: number = 0; i < boardSize * boardSize; i++) {
+                initializedSquares = [
+                    ...initializedSquares,
+                    { selected: false, id: i, selectedBy: '' },
+                ]; // creating a 1D array of squares with {selected: false, id: 0,...,boardSize}
+            }
+            return {
+                ...state,
+                boardSize: boardSize,
+                squares: initializedSquares,
+                created: true,
+            };
+        }
+        case REMOVE_BOARD: {
+            return initialBoardState; // return initialBoardState
+        }
+        case CLEAR_BOARD: { // kinda unclear how I should do this with {...state,...} and not Object.assign (though the end result should be the same)
+            return Object.assign(
```

Comment:
what wasn't working with spread operator?

Reply:
I need to look into how the spread operator works with deeper level elements, because for some reason I couldn't get it to work

Reply:
spread operator is shallow, so it spreads only one level. If you need to copy deeper levels, you will need to spread them separately. No reason you can't do nested spreads, eg.
{
  ...state,
  stuff: 'abc',
  squares: [...state.squares]
}

Reply:
but these days it's veeeery rare when you need to use Object.asign, unless you are working in an object oriented way (and here we are definitely not)